crates/typst-cli/src/download.rs

```diff
@@ -3,37 +3,89 @@
 // https://github.com/rust-lang/rustup/blob/master/src/cli/download_tracker.rs
 
 use std::collections::VecDeque;
+use std::error::Error;
+use std::fmt::Display;
 use std::io::{self, ErrorKind, Read, Write};
 use std::sync::Arc;
 use std::time::{Duration, Instant};
 
+use ecow::{eco_format, EcoString};
 use native_tls::{Certificate, TlsConnector};
-use once_cell::sync::Lazy;
+use once_cell::sync::OnceCell;
 use ureq::Response;
 
 use crate::terminal;
 
 /// Keep track of this many download speed samples.
 const SPEED_SAMPLES: usize = 5;
 
-/// Lazily loads a custom CA certificate if present, but if there's an error
-/// loading certificate, it just uses the default configuration.
-static CERT: Lazy<Option<Certificate>> = Lazy::new(|| {
-    let path = crate::ARGS.cert.as_ref()?;
-    let pem = std::fs::read(path).ok()?;
-    Certificate::from_pem(&pem).ok()
-});
+/// Load a certificate from the file system if the `--cert` argument or
+/// `TYPST_CERT` environment variable is present. The certificate is cached for
+/// efficiency.
+///
+/// - Returns `None` if `--cert` and `TYPST_CERT` are not set.
+/// - Returns `Some(Ok(cert))` if the certificate was loaded successfully.
+/// - Returns `Some(Err(err))` if an error occurred while loading the certificate.
+fn cert() -> Option<Result<&'static Certificate, EcoString>> {
+    static CERT: OnceCell<Certificate> = OnceCell::new();
+    crate::ARGS.cert.as_ref().map(|path| {
+        CERT.get_or_try_init(|| {
+            let pem = std::fs::read(path)
+                .map_err(|err| eco_format!("cannot read certificate: {err}"))?;
+            Certificate::from_pem(&pem)
+                .map_err(|err| eco_format!("cannot parse certificate: {err}"))
+        })
+    })
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub enum DownloadError {
```

Comment:
`ureq::Error` already implements `From<io::Error>` so perhaps we can just use a custom `io::Error` to pass through the certificate failures.

Reply:
Fixed. Thanks for the review!